File: solutions/234. Palindrome Linked List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* r = NULL;
    bool ans = true;
    
    void solve(ListNode* rt){
        if(rt == NULL){
            return ;
        }
        solve(rt->next);
        
        if(rt->val != r->val){
            ans = false;
            return;
        }
        r = r->next;
    }
    
    bool isPalindrome(ListNode* head) {
        // stack<int> cache;
        // ListNode* r = head;
        // while(r){
        //     cache.push(r->val);
        //     r = r->next;
        // }
        // r = head;
        // while(r){
        //     if(cache.top() != r->val){
        //         return false;
        //     }
        //     r = r->next;
        //     cache.pop();
        // }
        // return true;
        
        r = head;
        solve(head);
        return ans;
    }
};

File: ..solutions/234. Palindrome Linked List.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    private ListNode curr = null;
    public boolean isPalindrome(ListNode head) {
        curr = head;
        boolean result = _isPalindrome(head);
        return result;
    }
​
    private boolean _isPalindrome(ListNode head) {
​
        if (head == null) {
            return true;
        }
​
        // faith
        boolean rres = _isPalindrome(head.next);
        if (rres == false) return false;
​
        boolean my_status = head.val == curr.val;
        curr = curr.next;
​
        return my_status;
    }
}
